enhancements/machine-api/azure-ultra-disks.md

```diff
@@ -0,0 +1,261 @@
+
+### User Stories
+
+**Story 1**
+As a developer, I'd like to have the ability to use Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 2**
+As a developer, I'd like to specify some basic specs (Disk Size, Logical Unit Number (LUN), Caching Type) for creating Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 3**
+As a developer, I'd like to have the ability to create Persistent Volume Claims (PVCs) which can dynamically bind to a StorageClass backed by Azure Ultra Disk and to mount them successfully to Pods for my workloads needs.
+
+### API Extensions
+
+#### Extension for Ultra Disks as Data Disks
+
+The API design is taken from the upstream Cluster API Provider for Azure (CAPZ), [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/azuremachine_types.go#L77-L79) and [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/types.go#L465-L483).
+The upstream design is followed as closely as possible to ensure smoother future migrations between MAPI and CAPI.
+
+A new field will be added to the `AzureMachineProviderSpec` struct `DataDisks`. The interface will be a slice of `DataDisk` structs allowing multiple Data Disks and various fields specific to them to be further specified.
+
+A `DataDisk` has one Composite type field, `ManagedDisk` of type `*ManagedDiskParameters`, which is a type already defined in the Machine API spec, used for the `ManagedDisk` field in the `OSDisk` type struct, another `AzureMachineProviderSpec` field.
+
+
+```go
+type AzureMachineProviderSpec struct {
+  // Existing fields will not be modified
+  ...
+
+  // DataDisk specifies the parameters that are used to add one or more data disks to the machine
+  // +optional
+  DataDisks []DataDisk `json:"dataDisks,omitempty"`
+}
+
+// DataDisk specifies the parameters that are used to add one or more data disks to the machine.
+type DataDisk struct {
+  // NameSuffix is the suffix to be appended to the machine name to generate the disk name.
+  // Each disk name will be in format <machineName>_<nameSuffix>.
+  // The overall disk name must not exceed 80 chars in length and can only contain letters, numbers, underscores, periods or hyphens.
+  NameSuffix string `json:"nameSuffix"`
+  // DiskSizeGB is the size in GB to assign to the data disk.
+  // +kubebuilder:validation:Minimum=4
+  DiskSizeGB int32 `json:"diskSizeGB"`
+  // ManagedDisk specifies the Managed Disk parameters for the data disk.
+  // +optional
+  ManagedDisk *ManagedDiskParameters `json:"managedDisk,omitempty"`
```

Comment:
Does this need to be a pointer? What's the benefit here?

Reply:
Yes good point. Since I tried mirroring the upstream API design as close as possible, I kept the pointer as it was there for the upstream design.

But I agree with you it's not _necessary_ and we could avoid it since we don't have a specific reason to do so.
Also the `OSDisk` type API, which [also takes a `ManagedDiskParameters`](https://github.com/openshift/api/blob/b632c5fc10c0ea86cb18577db9388109a43d465f/machine/v1beta1/types_azureprovider.go#L203), doesn't reference it as a pointer, so another reason not to have it in there. I'll remove it.